src/v/cluster/cluster_utils.h

```diff
@@ -38,32 +39,37 @@ patch<broker_ptr> calculate_changed_brokers(
   const std::vector<broker_ptr>& new_list,
   const std::vector<broker_ptr>& old_list);
 
+class getTopicName {
+public:
+    getTopicName(){};
+    virtual double operator()(double left, double right) = 0;
+};
+
 /// Creates the same topic_result for all requests
 // clang-format off
 template<typename T>
-CONCEPT(requires requires(const T& req) {
-    { req.tp_ns } -> std::convertible_to<const model::topic_namespace&>;
-})
-// clang-format on
-std::vector<topic_result> create_topic_results(
-  const std::vector<T>& requests, errc error_code) {
+std::vector<topic_result>
+create_topic_results(const std::vector<T>& requests, errc error_code) {
     std::vector<topic_result> results;
     results.reserve(requests.size());
     std::transform(
       std::cbegin(requests),
       std::cend(requests),
       std::back_inserter(results),
-      [error_code](const T& r) { return topic_result(r.tp_ns, error_code); });
+      [error_code](const T& r) {
+          if constexpr (std::is_same_v<T, cluster::non_replicable_topic>)
+              return topic_result(r.name, error_code);
+          else if constexpr (std::is_same_v<T, cluster::topic_configuration> ||
+                   std::is_same_v<T, cluster::topic_properties_update>)
+              return topic_result(r.tp_ns, error_code);
+          else if constexpr (std::is_same_v<T, model::topic_namespace>)
+              return topic_result(r, error_code);
+          else if constexpr (std::is_same_v<T, cluster::custom_assignable_topic_configuration>)
+              return topic_result(r.cfg.tp_ns, error_code);
```

Comment:
consider adding an `else` case so that you can catch unused cases. here is an example: https://github.com/vectorizedio/redpanda/blob/dev/src/v/security/acl.h#L66 or you can also compose a static_assert of disjunctions using `is_same_v`

Reply:
Or the method could accept a future which performs the transformation and there would be no possibility of an unused case, IMO I think thats nicer then a monolithic function which performs the transformation based on type

Reply:
Rob's suggestion is good

Reply:
> 

I'm not sure how to do that, does that mean every time create_topic_results is invoked, a lambda function should be created and passed as parameter to create_topic_results ?

Reply:
I'm not sure how to do that, does that mean every time create_topic_results is invoked, a lambda function should be created and passed as parameter to create_topic_results ?

Reply:
I thought Rob's suggestion made sense, but I didn't think too hard about it. @graphcareful do you have a quick example?

Reply:
Yup like this:
```
auto new_rs = create_topic_results(requests, [e](const T& r) { return topic_result(r.name, e); });
```